### ISSUE TITLE
How to generate tracking mask

### ISSUE THREAD

**Issue by xduzhangjiayu** [tau-yihouxiang/EX-4D#14]
Hi Author,
Really awesome work! And I want to ask if the project contains the code of tracking mask generating? 
Thanks!

**Comment by mayuelala** [comment]
The tracking mask is in the training stage, I think this rep doesn't have this code. Could the authors provide the details code about this?

**Comment by tau-yihouxiang** [comment]
Hi, thank you for following, @xduzhangjiayu and @mayuelala. Below is the code for online generation of tracking mask videos.
```
cotracker = torch.hub.load("facebookresearch/co-tracker", "cotracker3_offline")
grid_size = random.randint(10, 50)
mask = generate_tracking_mask(video.clone(), self.cotracker, grid_size)
def generate_tracking_mask(frames, cotracker, grid_size=32):
    """
    Generate a mask video using CoTracker without any for loops
    
    Args:
        frames: Original video frames (torch tensor) with shape [B, C, T, H, W]
        cotracker: CoTracker model
        grid_size: Size of the tracking grid (default: 32)
        
    Returns:
        mask_video: Mask video as torch tensor [B, 1, T, H, W]
    """
    frames = (frames + 1) / 2 * 255.0  # Convert to [0, 255]
    frames = frames.permute(0, 2, 1, 3, 4)  # B, T, C, H, W
    device = frames.device
    B, T, C, H, W = frames.shape

    # Run CoTracker to get tracks and visibility
    with torch.no_grad():
        pred_tracks, pred_visibility = cotracker(frames, grid_size=grid_size)
        torch.cuda.empty_cache()  # Clear GPU memory if using CUDA

    # Extract tracking information
    B, T, N, _ = pred_tracks.shape
    
    # Convert visibility to boolean
    visibility = pred_visibility.squeeze(-1).bool()  # [B, T, N]
    
    # Calculate mask size
    random_scale = 1.0
    mask_width = W // (grid_size - 1) * random_scale
    mask_height = H // (grid_size - 1) * random_scale
    half_w = mask_width // 2
    half_h = mask_height // 2
    
    # Create mask tensor
    mask_video = torch.zeros((B, 1, T, H, W), dtype=torch.float32, device=device)
    
    # Set first frame to all ones
    # mask_video[:, 0, 0] = 1.0
    
    # Create coordinate grids for each pixel in the image
    y_grid, x_grid = torch.meshgrid(
        torch.arange(H, device=device),
        torch.arange(W, device=device),
        indexing='ij'
    )

    # fix_invisible_index = torch.randperm(grid_size**2)[:random.randint(0, grid_size**2//2)]
    # fix_invisible_index = fix_invisible_index.sort()[0]
    # visibility[..., fix_invisible_index] = False
    
    # Handle frames after the first one (t=0)
    # Process all time steps at once for all batches
    for t in range(0, T):
        # Get points for this time step
        points = pred_tracks[:, t]  # [B, N, 2]
        vis = visibility[:, t]  # [B, N]
        
        # Create binary mask for this frame
        frame_mask = torch.zeros((B, H, W), device=device)
        
        # Create a mask where each pixel gets set to 1 if it's within the 
        # mask_width × mask_height area of any tracked point
        x_coords = points[:, :, 0].unsqueeze(-1).unsqueeze(-1)  # [B, N, 1, 1]
        y_coords = points[:, :, 1].unsqueeze(-1).unsqueeze(-1)  # [B, N, 1, 1]
        
        # Calculate distance from each pixel to each point
        x_dist = torch.abs(x_grid.unsqueeze(0).unsqueeze(1) - x_coords)  # [B, N, H, W]
        y_dist = torch.abs(y_grid.unsqueeze(0).unsqueeze(1) - y_coords)  # [B, N, H, W]
        
        # Create mask for points within the area
        within_x = x_dist <= half_w
        within_y = y_dist <= half_h
        point_mask = (within_x & within_y).float()  # [B, N, H, W]
        
        # Apply visibility mask
        point_mask = point_mask * vis.unsqueeze(-1).unsqueeze(-1).float()  # [B, N, H, W]
        
        # Combine masks (max along N dimension)
        frame_mask = torch.max(point_mask, dim=1)[0]  # [B, H, W]
        
        # Update mask video
        mask_video[:, 0, t] = frame_mask
    
    for b in range(B):
        # Generate random odd kernel size between 1 and 29
        morpho_kernel = random.randrange(11, 30, 2)
        morpho_kernel = np.ones((morpho_kernel, morpho_kernel), np.uint8)
        for t in range(T):
            mask = (mask_video[b, 0, t].cpu().numpy() * 255).astype(np.uint8)
            mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, morpho_kernel) 
            mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, morpho_kernel) 

            erode_kernel = random.randint(2, 8)
            mask = cv2.erode(mask, np.ones((erode_kernel, erode_kernel), np.uint8), iterations=1)
            
            _, mask = cv2.threshold(mask, 127, 255, cv2.THRESH_BINARY)
            mask = mask.astype(np.float32) / 255.0
            mask_video[b, 0, t] = torch.from_numpy(mask).to(device)
    
    mask_video = mask_video.expand(-1, 3, -1, -1, -1)  # Expand to [B, 3, T, H, W]
    return mask_video
```

**Comment by xduzhangjiayu** [comment]
@tau-yihouxiang Thanks so much for the code~
